regression - nightly 2023-03-30 -> 2023-03-31 causes linker error when importing wasm_bindgen::JsValue in a proc_macro

<!--
Thank you for filing a regression report! 🐛 A regression is something that changed between versions of Rust but was not supposed to.

Please provide a short summary of the regression, along with any information you feel is relevant to replicate it.
-->

### Notes for other investigators

#### (2023-07-19)

1. stylist-rs has been updated with a workaround, so the reproduction described below will need adapting (e.g. check out that repo from around https://github.com/futursolo/stylist-rs/commit/fbd788f0a9a63ccadf6a25e028829a25224140eb or earlier, since that will not include the workaround added in https://github.com/futursolo/stylist-rs/pull/123 )

2. This problem exhibits itself only with certain numbers of codegen-units. The default setting for codegen-units is host dependent (the default is chosen, in part, as a function of the number of cores the host has available, in order to parallelize the code-generation performed by LLVM). In my own experiments, the problem is masked (does not arise) for codegen-units={1,2,4,8}, and the problem *does* arise for codegen-units={16,32,64,128,256}. You are best off passing `-Ccodegen-units` explicit when investigating this, just to ward against that being an invisible factor in the experiments.

(bug report follows below)

### Code

I tried this code:

```toml
[package]
name = "nightly_stylist"
version = "0.1.0"
edition = "2021"

[dependencies]
stylist-macros = "0.12"
```

```rust
fn main() {}
```

[stylist-macros](https://docs.rs/stylist-macros/0.12.0/stylist_macros/)

I expected to see this happen: `cargo check` succeeds in debug mode.

Instead, this happened: `cargo check` fails in debug mode.

```console
  = note: /usr/bin/ld: __wbindgen_realloc: undefined version: 
          /usr/bin/ld: __wbindgen_malloc: undefined version: 
          /usr/bin/ld: __wbindgen_free: undefined version: 
          /usr/bin/ld: __wbindgen_exn_store: undefined version: 
          /usr/bin/ld: failed to set dynamic section sizes: bad value
          collect2: error: ld returned 1 exit status
```

### Version it worked on

```
rustc 1.70.0-nightly (ec2f40c6b 2023-03-30)
binary: rustc
commit-hash: ec2f40c6b04f0e9850dd1f454e8639d319f4ed9b
commit-date: 2023-03-30
host: x86_64-unknown-linux-gnu
release: 1.70.0-nightly
LLVM version: 16.0.0
```

### Version with regression

```
rustc 1.70.0-nightly (5e1d3299a 2023-03-31)
binary: rustc
commit-hash: 5e1d3299a290026b85787bc9c7e72bcc53ac283f
commit-date: 2023-03-31
host: x86_64-unknown-linux-gnu
release: 1.70.0-nightly
LLVM version: 16.0.0
```

### Cargo bisect-rustc

```
searched toolchains ec2f40c6b04f0e9850dd1f454e8639d319f4ed9b through 5e1d3299a290026b85787bc9c7e72bcc53ac283f


********************************************************************************
Regression in 22a7a19f9333bc1fcba97ce444a3515cb5fb33e6
********************************************************************************
```
*isn't immediately obvious why [this change](https://github.com/rust-lang/rust/commit/22a7a19f9333bc1fcba97ce444a3515cb5fb33e6) would cause the problem

searched nightlies: from nightly-2023-03-27 to nightly-2023-04-02
regressed nightly: nightly-2023-04-01
searched commit range: https://github.com/rust-lang/rust/compare/ec2f40c6b04f0e9850dd1f454e8639d319f4ed9b...5e1d3299a290026b85787bc9c7e72bcc53ac283f
regressed commit: https://github.com/rust-lang/rust/commit/22a7a19f9333bc1fcba97ce444a3515cb5fb33e6

<details>
<summary>bisected with <a href='https://github.com/rust-lang/cargo-bisect-rustc'>cargo-bisect-rustc</a> v0.6.6</summary>


Host triple: x86_64-unknown-linux-gnu
Reproduce with:
```bash
cargo bisect-rustc --start 2023-03-27 --end 2023-04-02 
```
</details>

<!--
If you know when this regression occurred, please add a line like below, replacing `{channel}` with one of stable, beta, or nightly.
@rustbot modify labels: +regression-from-nightly-to-nightly -regression-untriaged
-->

--- a/compiler/rustc_codegen_ssa/src/back/linker.rs
+++ b/compiler/rustc_codegen_ssa/src/back/linker.rs
@@ -13,6 +13,7 @@ use std::{env, mem, str};
 use rustc_hir::def_id::{CrateNum, LOCAL_CRATE};
 use rustc_metadata::find_native_static_library;
 use rustc_middle::middle::dependency_format::Linkage;
+use rustc_middle::middle::exported_symbols;
 use rustc_middle::middle::exported_symbols::{ExportedSymbol, SymbolExportInfo, SymbolExportKind};
 use rustc_middle::ty::TyCtxt;
 use rustc_session::config::{self, CrateType, DebugInfo, LinkerPluginLto, Lto, OptLevel, Strip};
@@ -658,8 +659,6 @@ impl<'a> Linker for GccLinker<'a> {
         if !self.sess.target.limit_rdylib_exports {
             return;
         }
-
-        // FIXME(#99978) hide #[no_mangle] symbols for proc-macros
 
         let is_windows = self.sess.target.is_like_windows;
         let path = tmpdir.join(if is_windows { "list.def" } else { "list" });
@@ -1679,8 +1678,15 @@ pub(crate) fn exported_symbols(tcx: TyCtxt<'_>, crate_type: CrateType) -> Vec<St
         return exports.iter().map(ToString::to_string).collect();
     }
 
-    let mut symbols = Vec::new();
+    if let CrateType::ProcMacro = crate_type {
+        exported_symbols_for_proc_macro_crate(tcx)
+    } else {
+        exported_symbols_for_non_proc_macro(tcx, crate_type)
+    }
+}
 
+fn exported_symbols_for_non_proc_macro(tcx: TyCtxt<'_>, crate_type: CrateType) -> Vec<String> {
+    let mut symbols = Vec::new();
     let export_threshold = symbol_export::crates_export_threshold(&[crate_type]);
     for_each_exported_symbols_include_dep(tcx, crate_type, |symbol, info, cnum| {
         if info.level.is_below_threshold(export_threshold) {
@@ -1689,6 +1695,26 @@ pub(crate) fn exported_symbols(tcx: TyCtxt<'_>, crate_type: CrateType) -> Vec<St
     });
 
     symbols
+}
+
+fn exported_symbols_for_proc_macro_crate(tcx: TyCtxt<'_>) -> Vec<String> {
+    let mut symbols = Vec::new();
+
+    let stable_crate_id = tcx.sess.local_stable_crate_id();
+    let proc_macro_decls_name = tcx.sess.generate_proc_macro_decls_symbol(stable_crate_id);
+    let metadata_symbol_name = exported_symbols::metadata_symbol_name(tcx);
+
+    // You would think that both the two names would always be there, but in
+    // pnkfelix's local experiments that was not case. So instead we walk the
+    // list and only add them if they *are* there.
+    for_each_exported_symbols_include_dep(tcx, CrateType::ProcMacro, |symbol, _info, cnum| {
+        let name = symbol_export::symbol_name_for_instance_in_crate(tcx, symbol, cnum);
+        if name == proc_macro_decls_name || name == metadata_symbol_name {
+            symbols.push(name);
+        }
+    });
+
+    return symbols;
 }
 
 pub(crate) fn linked_symbols(

--- a/compiler/rustc_codegen_ssa/src/back/linker.rs
+++ b/compiler/rustc_codegen_ssa/src/back/linker.rs
@@ -1698,23 +1698,16 @@ fn exported_symbols_for_non_proc_macro(tcx: TyCtxt<'_>, crate_type: CrateType) -
 }
 
 fn exported_symbols_for_proc_macro_crate(tcx: TyCtxt<'_>) -> Vec<String> {
-    let mut symbols = Vec::new();
+    // `exported_symbols` will be empty when !should_codegen.
+    if !tcx.sess.opts.output_types.should_codegen() {
+        return Vec::new();
+    }
 
     let stable_crate_id = tcx.sess.local_stable_crate_id();
     let proc_macro_decls_name = tcx.sess.generate_proc_macro_decls_symbol(stable_crate_id);
     let metadata_symbol_name = exported_symbols::metadata_symbol_name(tcx);
 
-    // You would think that both the two names would always be there, but in
-    // pnkfelix's local experiments that was not case. So instead we walk the
-    // list and only add them if they *are* there.
-    for_each_exported_symbols_include_dep(tcx, CrateType::ProcMacro, |symbol, _info, cnum| {
-        let name = symbol_export::symbol_name_for_instance_in_crate(tcx, symbol, cnum);
-        if name == proc_macro_decls_name || name == metadata_symbol_name {
-            symbols.push(name);
-        }
-    });
-
-    return symbols;
+    vec![proc_macro_decls_name, metadata_symbol_name]
 }
 
 pub(crate) fn linked_symbols(

--- a/tests/ui/proc-macro/auxiliary/exports_no_mangle.rs
+++ b/tests/ui/proc-macro/auxiliary/exports_no_mangle.rs
@@ -1,0 +1,9 @@
+#![crate_type="lib"]
+
+// Issue 111888: this crate (1.) is imported by a proc-macro crate and (2.)
+// exports a no_mangle function; that combination of acts was broken for some
+// period of time. See further discussion in the test file that imports this
+// crate.
+
+#[no_mangle]
+pub fn some_no_mangle_function() { }

--- a/tests/ui/proc-macro/auxiliary/exports_no_mangle.rs
+++ b/tests/ui/proc-macro/auxiliary/exports_no_mangle.rs
@@ -1,3 +1,5 @@
+// force-host
+// no-prefer-dynamic
 #![crate_type="lib"]
 
 // Issue 111888: this crate (1.) is imported by a proc-macro crate and (2.)

--- a/tests/ui/proc-macro/no-mangle-in-proc-macro-issue-111888.rs
+++ b/tests/ui/proc-macro/no-mangle-in-proc-macro-issue-111888.rs
@@ -1,0 +1,20 @@
+// run-pass
+// aux-build:exports_no_mangle.rs
+#![crate_type = "proc-macro"]
+
+// Issue #111888: this proc-macro crate imports another crate that itself
+// exports a no_mangle function.
+//
+// That combination was broken for a period of time, because:
+//
+// In PR #99944 we *stopped* exporting no_mangle symbols from
+// proc-macro crates. The constructed linker version script still referred
+// to them, but resolving that discrepancy was left as a FIXME in the code.
+//
+// In PR #108017 we started telling the linker to check (via the
+// `--no-undefined-version` linker invocation flag) that every symbol referenced
+// in the "linker version script" is actually present in the linker input. So
+// the unresolved discrepancy from #99944 started surfacing as a compile-time
+// error.
+
+extern crate exports_no_mangle;

--- a/tests/ui/proc-macro/no-mangle-in-proc-macro-issue-111888.rs
+++ b/tests/ui/proc-macro/no-mangle-in-proc-macro-issue-111888.rs
@@ -1,4 +1,4 @@
-// run-pass
+// build-pass
 // aux-build:exports_no_mangle.rs
 #![crate_type = "proc-macro"]
 

--- a/tests/ui/proc-macro/no-mangle-in-proc-macro-issue-111888.rs
+++ b/tests/ui/proc-macro/no-mangle-in-proc-macro-issue-111888.rs
@@ -1,4 +1,6 @@
 // build-pass
+// force-host
+// no-prefer-dynamic
 // aux-build:exports_no_mangle.rs
 #![crate_type = "proc-macro"]
 

